### PR TITLE
Introduce new `suspend` #891

### PR DESCRIPTION
**Pull request by kamilkloch**
Added `suspendTotal` and `suspendTotalWith`, as discussed in #891. WIP.

### CODE REVIEW

**Comment by jdegoes** [comment]
@kamilkloch It's not quite enough to use the types recommended in #891. Rather, if these are primitive operations baked into the interpreter, then we need to try / catch around them as done for `ZIO.effect`.

I recommend we motivate this by adding some tests. In particular, test that:

 - `ZIO.effectSuspend(throw TestError).either.flatMap(_ must_=== TestError)`
 - `ZIO.effectSuspendWith(_ => throw TestError).either.flatMap(_ must_=== TestError)`

**Comment by kamilkloch** [comment]
I am not sure if the signatures are right (with regard to the error type). Current
```scala
effectSuspend(zio: => ZIO[R, E, A]): ZIO[R, E, A]
```
is as generic as possible, but `effectSuspend`handles `throwables`, not generic `E`s. 

On the other hand, 
```scala
effectSuspend[R, A](task: => TaskR[R, A]): TaskR[R, A]
```
does not work for current `FunctionIO.Impure` implemenation (generic `E`):
```scala
private[zio] final class Impure[E, A, B](val apply0: A => B) extends FunctionIO[E, A, B] {
    val run: A => IO[E, B] = a =>
      IO.suspend {
        try IO.succeed[B](apply0(a))
        catch {
          case e: FunctionIOError[_] => IO.fail[E](e.unsafeCoerce[E])
        }
      }
  }
```

File: core/shared/src/main/scala/zio/Queue.scala [hunk 1]
```diff
@@ -132,7 +132,7 @@ object Queue {
       }
 
       final def handleSurplus(as: List[A], queue: MutableConcurrentQueue[A]): UIO[Boolean] =
-        ZIO.suspend {
+        ZIO.effectSuspend {
```
Comment (kamilkloch) [inline core/shared/src/main/scala/zio/Queue.scala]: effectSuspendTotal instead?

File: core/shared/src/main/scala/zio/Queue.scala [hunk 2]
```diff
@@ -228,7 +228,7 @@ object Queue {
     final def offer(a: A): UIO[Boolean] = offerAll(List(a))
 
     final def offerAll(as: Iterable[A]): UIO[Boolean] =
-      ZIO.suspend {
+      ZIO.effectSuspend {
```
Comment (kamilkloch) [inline core/shared/src/main/scala/zio/Queue.scala]: effectSuspendTotal instead?

File: core/shared/src/main/scala/zio/Queue.scala [hunk 3]
```diff
@@ -271,7 +271,7 @@ object Queue {
     final val isShutdown: UIO[Boolean] = shutdownHook.poll.map(_.isDefined)
 
     final val take: UIO[A] =
-      ZIO.suspend {
+      ZIO.effectSuspend {
```
Comment (kamilkloch) [inline core/shared/src/main/scala/zio/Queue.scala]: effectSuspendTotal instead?

File: core/shared/src/main/scala/zio/Queue.scala [hunk 4]
```diff
@@ -298,7 +298,7 @@ object Queue {
       }
 
     final val takeAll: UIO[List[A]] =
-      ZIO.suspend {
+      ZIO.effectSuspend {
```
Comment (kamilkloch) [inline core/shared/src/main/scala/zio/Queue.scala]: effectSuspendTotal instead?

File: core/shared/src/main/scala/zio/Queue.scala [hunk 5]
```diff
@@ -311,7 +311,7 @@ object Queue {
       }
 
     final def takeUpTo(max: Int): UIO[List[A]] =
-      ZIO.suspend {
+      ZIO.effectSuspend {
```
Comment (kamilkloch) [inline core/shared/src/main/scala/zio/Queue.scala]: effectSuspendTotal instead?

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 6]
```diff
@@ -1420,7 +1420,7 @@ private[zio] trait ZIOFunctions extends Serializable {
   final def runtime[R >: LowerR]: ZIO[R, Nothing, Runtime[R]] =
     for {
       environment <- environment[R]
-      platform    <- suspendWith(ZIO.succeed)
+      platform    <- effectSuspendWith(ZIO.succeed)
```
Comment (kamilkloch) [inline core/shared/src/main/scala/zio/ZIO.scala]: effectSuspendTotalWith instead?

File: core/shared/src/main/scala/zio/Queue.scala (hunk 5 above)
Comment (jdegoes) [reply core/shared/src/main/scala/zio/Queue.scala]: Yes, the `total` variants will _not_ catch exceptions and will return `UIO`.

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 6 above)
Comment (jdegoes) [reply core/shared/src/main/scala/zio/ZIO.scala]: Right on!

**Comment by jdegoes** [comment]
> I am not sure if the signatures are right (with regard to the error type). Current

I think we can have, as you suggested, `total` variants that do not translate exceptions into typed error values. These can be generic for any `E`. Anything that catches a `Throwable` (the partial / non-total variants) will have to fix to `Throwable`.

**Comment by kamilkloch** [comment]
> I think we can have, as you suggested, `total` variants that do not translate exceptions into typed error values. These can be generic for any `E`. Anything that catches a `Throwable` (the partial / non-total variants) will have to fix to `Throwable`.

Thanks. Let me re-iterate to make sure I get it right. Current state in the PR:
```scala
effectSuspend(zio: => ZIO[R, E, A]): ZIO[R, E, A]
effectSuspendTotal(zio: => ZIO[R, Nothing, A]): ZIO[R, Nothing, A]
```
Desired state:
```scala
effectSuspend(taskr: => TaskR[R, A]): TaskR[R, A]
effectSuspendTotal(zio: => ZIO[R, E, A]): ZIO[R, E, A]
```
?

**Comment by jdegoes** [comment]
@kamilkloch Yes, and remember the partial variations should catch `throw new Throwable` (e.g.) and allow you to `catchAll` that error, whereas the total versions should _not_ catch it.

Finally, remember the `with` variations that pass the `Platform` for low-level stuff...

**Comment by kamilkloch** [comment]
`effectSuspendTotal` variations:
 - ZIO
   ```scala
   def effectSuspendTotal[R, E, A](zio: => ZIO[R, E, A]): ZIO[R, E, A]
   ```
 - RIO
   ```scala
   def effectSuspendTotal[R, A](rio: => RIO[R, A]): RIO[R, A]=> rio)
   ```
 - UIO
   ```scala
   def effectSuspendTotal[A](uio: => UIO[A]): UIO[A]
   ```
 - Task
    ```scala
    def effectSuspendTotal[A](task: => Task[A]): Task[A]
   ```

Only available in ZIO, RIO, Task:
```scala
def effectSuspend[R, A](rio: => RIO[R, A]): RIO[R, A]
```

**Comment by kamilkloch** [comment]
> Don't forget the `With` variants that pass in the `Platform`.

It's already there :)

> And we need tests that:
> 
> * `*Total` variants cannot be "caught" with `catchAll`
> * _Partial_ variant (`effectSuspend` / `effectSuspendWith`) **can** be "caught" with `catchAll`

Test for `partial` was there (`unsafeRun(ZIO.effectSuspend[Any, Nothing](throw ExampleError).either) must_=== Left(ExampleError)`). I just added the missing test for the `total` variant.

**Comment by kamilkloch** [comment]
@jdegoes Do we miss anything? :)

**Comment by kamilkloch** [comment]
@ghostdogpr @jdegoes methods are now sorted.

**Comment by jdegoes** [comment]
giphy

**Pull request merged**